### ISSUE TITLE
Support some of the open source local LLMs such as WizardCoder15B

### ISSUE THREAD

**Issue by davidliudev** [getcursor/cursor#668]
**Is your feature request related to a problem? Please describe.**

My company unfortunately cannot benefit from the AI power-tools because anything that involves code remote upload will be violating our internal policy. 

We need not only a local mode but a local LLM too so we can ensure that nothing leaves the local environment while leveraging the AI.

**Describe the solution you'd like**
We now have very powerful (and continue evolving) local LLMs for coding such as StarCoder as well as the recent star WizardCoder which is said to exceed SOTA in various benchmarks. It is no doubt that this will continue evolve. Why not support these local models in cursor.so?

**Additional context**
Text generation webUI and llama.cpp/kobold.cpp does have api support so we could potentially and hopefully integrate this with relative ease

To summarize, what we (and probably a lot of people) need is:

1. 100% local mode. No remote server at all
2. Local and swappable open source LLMs

cursor.so is a great editor, especially the inline diff. Hope the team can make it even better by supporting more open source LLMs.

**Comment by truell20** [comment]
Thanks for writing in! In our experience tinkering with OSS models, they perform significantly worse than GPT-4. At the moment, our focus is on building the best software development experience with AI. That focus will unfortunately mean we aren't a fit for some companies/individuals when it comes to locality and cost.

Appreciate the suggestion. I just want to be honest about what we will and won't support up front. 

Would be curious to hear if you think we're off base in our estimation of these OSS models (especially curious to hear about first hand experience in lieu of benchmarks).

**Comment by Agentsmithforge** [comment]
@truell20 i believe you are 100% off base with this estimation, and i can prove it!
i can try implementing this myself if you could give me some pointers as to where in the code should i look for the functions calling the openAI api.

**Comment by JohnPeng47** [comment]
Guys you understand they won't support because of they do, that means they expose their prompts + code context right?

**Comment by davidliudev** [comment]
As Visual Studio Code has official move into the domain of AI copilot and chat-related feature, and considering their resources and money, IMO the only opportunity for AI editor startups is to address some of the concerns that the Big Company won't, such as locality and data security.

Locality and open source is very important to many teams and companies as many folks I know have yet adopted AI tools due to their strict company policy. I do understand that your team has your own priorities and roadmaps but if possible please seriously consider such directions as this is truly what the community need.

cc @truell20

**Comment by atljoseph** [comment]
Please stop framing it from the standpoint of, “it would be a bad experience”. Let the people that would be using it decide that. And if they can’t use it then they can’t decide it.

**Comment by cashgarman** [comment]
There are tutorials out there to get open-source local models working, have a quick search if you are interested. But please understand, I'm sure a ton of development went into Cursor, and it shows in it's extremely high quality UX and features, and their business model is subscription-based. So for them to actively support the use of local LLMs is, understandably, not in their best interests. I, personally, would support them at $50/month for full local LLM support, but I think most people would use local LLMs to bypass their only revenue stream.

**Comment by AllyourBaseBelongToUs** [comment]
> There are tutorials out there to get open-source local models working, have a quick search if you are interested. But please understand, I'm sure a ton of development went into Cursor, and it shows in it's extremely high quality UX and features, and their business model is subscription-based. So for them to actively support the use of local LLMs is, understandably, not in their best interests. I, personally, would support them at $50/month for full local LLM support, but I think most people would use local LLMs to bypass their only revenue stream.

All my searches so far came empty, besides using Open router, which isn't local 

Are you able to point to some resources ?

Llamaserver would be nice

**Comment by davidliudev** [comment]
It has been a good year since my first post and my original position still stands. I personally liked Cursor a lot and I have used it a lot for my own personal projects.

However for my team it is a different story. Everyone knows that AI is good for them and it will accelerate speed and efficiency but I can't get my team equipped with these tools due to the cloud nature of them, which means all the context, including private keys and other secrets, even if only stored in a .env file which is git-ignored and stayed local for testing, will be sent to some remote server. And our internal policy cannot allow this to happen.

For the past year the local LLM ability has greatly improved, compared to last year. We've seen quite good quality local LLMs such as Gemma2, Phi3.5, Llama3.1 and Deepseek Coder and more, which now makes it even more powerful than before.

And no, it is not because we want to bypass the revenue model. We are willing to pay for a subscription or whatever business model that is suitable as long as this stays local.

Currently some of our team members are experimenting with Jetbrain's local AI copilots. Many find it helpful.

**Comment by davidliudev** [comment]
After revisiting this and search online for a while I guess now we can do something similar to a local LLM using an override baseURL for a local llm server with OpenAI compatible APIs? Will give it a try later

**Comment by jarmitage** [comment]
> Guys you understand they won't support because of they do, that means they expose their prompts + code context right?

This, and I think we can also assume that as Open AI is their investor, open LLMs goes against their pitch deck / business plan.

**Comment by tylerkatz** [comment]
> Thanks for writing in! In our experience tinkering with OSS models, they perform significantly worse than GPT-4. At the moment, our focus is on building the best software development experience with AI. That focus will unfortunately mean we aren't a fit for some companies/individuals when it comes to locality and cost.
> 
> Appreciate the suggestion. I just want to be honest about what we will and won't support up front.
> 
> Would be curious to hear if you think we're off base in our estimation of these OSS models (especially curious to hear about first hand experience in lieu of benchmarks).

Guys, they make their money by people paying for fast requests to cloud models.  They're not going to support local models where there is no concept of fast requests lol.  You're barking up the wrong tree.

**Comment by LeulAria** [comment]
So you can't use cursor with local llm for free, alright im unsubscribing let me explore other things out their.
